python/ray/serve/_private/replica.py

```diff
@@ -812,145 +809,203 @@ async def call_reconfigure(self, user_config: Any):
+                        r = (request_metadata.grpc_context, r.SerializeToString())
+                    await generator_result_callback(r)
+            elif result_is_async_gen:
+                async for r in result:
+                    if request_metadata.is_grpc_request:
+                        r = (request_metadata.grpc_context, r.SerializeToString())
+                    await generator_result_callback(r)
+            elif request_metadata.is_http_request and not is_asgi_app:
+                # For the FastAPI codepath, the response has already been sent over
+                # ASGI, but for the vanilla deployment codepath we need to send it.
+                await self._send_user_result_over_asgi(result, asgi_args)
+            elif not request_metadata.is_http_request:
+                # HTTP requests are always streaming but the method itself does not
+                # return a generator. Instead, the method is provided the result queue
+                # as its ASGI `send` interface to stream back results.
                 raise TypeError(
-                    f"Method '{runner_method.__name__}' returned a generator. You "
-                    "must use `handle.options(stream=True)` to call "
-                    "generators on a deployment."
+                    f"Called method '{user_method_name}' with "
+                    "`handle.options(stream=True)` but it did not return a "
+                    "generator."
                 )
-
-        except Exception as e:
-            function_name = "unknown"
-            if runner_method is not None:
-                function_name = runner_method.__name__
-            e = wrap_to_ray_error(function_name, e)
-            if request_metadata.is_http_request:
-                result = starlette.responses.Response(
-                    f"Unexpected error, traceback: {e}.", status_code=500
+        else:
+            if result_is_gen or result_is_async_gen:
+                raise TypeError(
+                    f"Method '{user_method_name}' returned a generator. "
+                    "You must use `handle.options(stream=True)` to call "
+                    "generators on a deployment."
                 )
-                await self._send_user_result_over_asgi(result, scope, receive, send)
-
-            raise e from None
-
-        if request_metadata.is_http_request and not isinstance(
-            self._callable, ASGIAppReplicaWrapper
-        ):
-            # For the FastAPI codepath, the response has already been sent over the
-            # ASGI interface, but for the vanilla deployment codepath we need to
-            # send it.
-            await self._send_user_result_over_asgi(result, scope, receive, send)
-        elif request_metadata.is_grpc_request:
-            result = (request_metadata.grpc_context, result.SerializeToString())
+            if request_metadata.is_grpc_request:
+                result = (request_metadata.grpc_context, result.SerializeToString())
```

Comment:
Does this else block not have a block that does `_send_user_result_over_asgi` because if the request is non-streaming, it must be grpc or over a handle (and not http)? (maybe a comment would help, although the logic can be followed from all the conditionals, it can be difficult to parse if you're not familiar with all the cases)

Reply:
Yes that's correct. Let me add a comment

Reply:
added an assertion to make it clear that HTTP requests don't fall through this section